### ISSUE TITLE
🚀 [syntax highlighting] improved language mappings via shebang

### ISSUE THREAD

**Issue by Kr1ss-XD** [dandavison/delta#485]
Some tools (especially `bat`) check source files for a shebang line and if present use this to assign the according syntax rules. I'm wondering if this would be possible for `delta`, too.

Currently, a source file seems to be considered a specific language only if its name/extension can be mapped to a known language. Therefore, generically named files (e.g. executable shell scripts without `*.sh` filename extension) are not being syntax highlighted by `delta`. Considering a shebang could be an option in addition to filenames or `[--map-syntax](https://github.com/dandavison/delta/issues/326)`.

Maybe it's even possible to use the algorithm/regexes which `bat` has already implemented ?

**Comment by dandavison** [comment]
Hi @Kr1ss-XD, the core issue here is that bat has access to the entire file, whereas delta (in its current form) only has access to the section of the file that happens to be in the diff hunk. It would be possible to change delta so that it (optionally) tries to find the file on disk (or from the git repo via libgit2).  I have wondered from the beginning whether we would want to do that. Of course, the file might not even exist, since delta simply accepts whatever diff is given to it on stdin, which could be entirely fictional.

**Comment by Kr1ss-XD** [comment]
Right, I'm aware that it's not as simple for `delta` as for `bat` which is given a filename as argument most of the time.

Since `delta` can recognize filenames in some cases, I wondered it could utilize these to open the file and check its contents.

> Of course, the file might not even exist, since delta simply accepts whatever diff is given to it on stdin, which could be entirely fictional.

I haven't considered this though.

**Comment by dandavison** [comment]
> Since delta can recognize filenames in some cases, I wondered if it could utilize these to open the file and check its contents.

Yes, I agree, this would be possible. And as you say, for things like executable shell scripts, I think it's the only way forwards.

**Comment by zachriggle** [comment]
This would be really nice.

I have a lot of e.g. Python scripts without the `.py` extension, and having a colorized Python-syntax-highlighted diff for these and this would be a game-changer.

**Comment by zachriggle** [comment]
@dandavison I created a simple solution to this issue, you should be able to use `file` on files without extension, and the above mappings, to automatically determine the syntax of a given file.

**Comment by dandavison** [comment]
Thanks @zachriggle. One thing we should check before proceeding is whether there is a rust crate that already does this and looks to be reliable. Let me know if you're aware of anything.

**Comment by flxai** [comment]
@dandavison There is [syntect](https://github.com/trishume/syntect) with a relevant function [`find_syntax_by_first_line`](https://docs.rs/syntect/latest/syntect/parsing/struct.SyntaxSet.html#method.find_syntax_by_first_line). Thanks to @jplatte for pointing this out in a private conversation.
